vi.spyOn behaving unexpectedly in Browser Mode

### Describe the bug

I was trying to migrate to Browser Mode, but I'm scratching my head over one test suite.

Scenario: testing a module that imports a variable from another file, and mocking this variable using vi.spyOn as recommended in https://vitest.dev/guide/mocking.html#mock-exported-variables.

In Node.js, this behaves as expected.

In Browser Mode, this crashes with:

```
Caused by: TypeError: Cannot redefine property: MODE
```

### Reproduction

https://stackblitz.com/edit/vitest-dev-vitest-qe32xcwn?file=package.json

### System Info

```shell
System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 20.19.1 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.8.2 - /usr/local/bin/npm
    pnpm: 8.15.6 - /usr/local/bin/pnpm
  npmPackages:
    @vitest/browser: ^3.2.3 => 3.2.3 
    @vitest/ui: latest => 3.2.3 
    playwright: ^1.53.0 => 1.53.0 
    vite: latest => 6.3.5 
    vitest: latest => 3.2.3
```

### Used Package Manager

npm

### Validations

- [x] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [x] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [x] Read the [docs](https://vitest.dev/guide/).
- [x] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [x] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [x] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/docs/guide/browser/index.md
+++ b/docs/guide/browser/index.md
@@ -584,3 +584,45 @@ test('renders a message', async () => {
 When using Vitest Browser, it's important to note that thread blocking dialogs like `alert` or `confirm` cannot be used natively. This is because they block the web page, which means Vitest cannot continue communicating with the page, causing the execution to hang.
 
 In such situations, Vitest provides default mocks with default returned values for these APIs. This ensures that if the user accidentally uses synchronous popup web APIs, the execution would not hang. However, it's still recommended for the user to mock these web APIs for better experience. Read more in [Mocking](/guide/mocking).
+
+### Spying on Module Exports
+
+Browser Mode uses the browser's native ESM support to serve modules. The module namespace object is sealed and can't be reconfigured, unlike in Node.js tests where Vitest can patch the Module Runner. This means you can't call `vi.spyOn` on an imported object:
+
+```ts
+import { vi } from 'vitest'
+import * as module from './module.js'
+
+vi.spyOn(module, 'method') // ❌ throws an error
+```
+
+To bypass this limitation, Vitest supports `{ spy: true }` option in `vi.mock('./module.js')`. This will automatically spy on every export in the module without replacing them with fake ones.
+
+```ts
+import { vi } from 'vitest'
+import * as module from './module.js'
+
+vi.mock('./module.js', { spy: true })
+
+vi.mocked(module.method).mockImplementation(() => {
+  // ...
+})
+```
+
+However, the only way to mock exported _variables_ is to export a method that will change the internal value:
+
+::: code-group
+```js [module.js]
+export let MODE = 'test'
+export function changeMode(newMode) {
+  MODE = newMode
+}
+```
+```js [module.test.ts]
+import { expect } from 'vitest'
+import { changeMode, MODE } from './module.js'
+
+changeMode('production')
+expect(MODE).toBe('production')
+```
+:::

--- a/eslint.config.js
+++ b/eslint.config.js
@@ -113,6 +113,7 @@ export default antfu(
       'unused-imports/no-unused-imports': 'off',
       'ts/method-signature-style': 'off',
       'no-self-compare': 'off',
+      'import/no-mutable-exports': 'off',
     },
   },
   {

--- a/packages/spy/src/index.ts
+++ b/packages/spy/src/index.ts
@@ -481,9 +481,11 @@ export function spyOn<T, K extends keyof T>(
   catch (error) {
     if (
       error instanceof TypeError
-      && error.message.includes('Cannot redefine property')
       && Symbol.toStringTag
       && (obj as any)[Symbol.toStringTag] === 'Module'
+      && (error.message.includes('Cannot redefine property')
+        || error.message.includes('Cannot replace module namespace')
+        || error.message.includes('can\'t redefine non-configurable property'))
     ) {
       throw new TypeError(
         `Cannot spy on export "${String(objMethod)}". Module namespace is not configurable in ESM. See: https://vitest.dev/guide/browser/#limitations`,

--- a/packages/spy/src/index.ts
+++ b/packages/spy/src/index.ts
@@ -467,14 +467,32 @@ export function spyOn<T, K extends keyof T>(
     state = fn.mock._state()
   }
 
-  const stub = tinyspy.internalSpyOn(obj, objMethod as any)
-  const spy = enhanceSpy(stub) as MockInstance
+  try {
+    const stub = tinyspy.internalSpyOn(obj, objMethod as any)
 
-  if (state) {
-    spy.mock._state(state)
+    const spy = enhanceSpy(stub) as MockInstance
+
+    if (state) {
+      spy.mock._state(state)
+    }
+
+    return spy
   }
+  catch (error) {
+    if (
+      error instanceof TypeError
+      && error.message.includes('Cannot redefine property')
+      && Symbol.toStringTag
+      && (obj as any)[Symbol.toStringTag] === 'Module'
+    ) {
+      throw new TypeError(
+        `Cannot spy on export "${String(objMethod)}". Module namespace is not configurable in ESM. See: https://vitest.dev/guide/browser/#limitations`,
+        { cause: error },
+      )
+    }
 
-  return spy
+    throw error
+  }
 }
 
 let callOrder = 0

--- a/test/browser/specs/runner.test.ts
+++ b/test/browser/specs/runner.test.ts
@@ -1,5 +1,6 @@
 import type { Vitest } from 'vitest/node'
 import type { JsonTestResults } from 'vitest/reporters'
+import { readdirSync } from 'node:fs'
 import { readFile } from 'node:fs/promises'
 import { beforeAll, describe, expect, onTestFailed, test } from 'vitest'
 import { rolldownVersion } from 'vitest/node'
@@ -68,10 +69,11 @@ describe('running browser tests', async () => {
     expect(vitest.projects.map(p => p.browser?.vite.config.optimizeDeps.entries))
       .toEqual(vitest.projects.map(() => expect.arrayContaining(testFiles)))
 
-    // This should match the number of actual tests from browser.json
-    // if you added new tests, these assertion will fail and you should
-    // update the numbers
-    expect(browserResultJson.testResults).toHaveLength(16 * instances.length)
+    const testFilesCount = readdirSync('./test')
+      .filter(n => n.includes('.test.'))
+      .length + 1 // 1 is in-source-test
+
+    expect(browserResultJson.testResults).toHaveLength(testFilesCount * instances.length)
     expect(passedTests).toHaveLength(browserResultJson.testResults.length)
     expect(failedTests).toHaveLength(0)
   })

--- a/test/browser/test/mocking.test.ts
+++ b/test/browser/test/mocking.test.ts
@@ -1,0 +1,18 @@
+import { expect, it, vi } from 'vitest'
+import * as module from '../src/calculator'
+
+it('spying on an esm module prints an error', () => {
+  const error: Error = (() => {
+    try {
+      vi.spyOn(module, 'calculator')
+      expect.unreachable()
+    }
+    catch (err) {
+      return err
+    }
+  })()
+  expect(error.name).toBe('TypeError')
+  expect(error.message).toMatchInlineSnapshot(`"Cannot spy on export "calculator". Module namespace is not configurable in ESM. See: https://vitest.dev/guide/browser/#limitations"`)
+
+  expect(error.cause).toBeInstanceOf(TypeError)
+})

